Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp

```diff
@@ -67,6 +67,91 @@ struct TextureMapperLayer::ComputeTransformData {
     }
 };
 
+class TextureMapperFlattenedLayer final {
+    WTF_MAKE_FAST_ALLOCATED;
+public:
+    explicit TextureMapperFlattenedLayer(const IntRect& rect, double zNear, double zFar)
+        : m_rect(rect)
+        , m_zNear(zNear)
+        , m_zFar(zFar)
+    { }
+
+    inline IntRect layerRect() const
+    {
+        return m_rect;
+    }
+
+    inline bool needsUpdate() const
+    {
+        return !m_textures.size();
+    }
+
+    void update(TextureMapperPaintOptions& options, const std::function<void(TextureMapperPaintOptions&)>& paintFunction)
+    {
+        double prevZNear, prevZFar;
+        options.textureMapper.getDepthRange(prevZNear, prevZFar);
+        options.textureMapper.setDepthRange(m_zNear, m_zFar);
+
+        m_textures.clear();
+
+        IntSize maxTextureSize = options.textureMapper.maxTextureSize();
+        tileLoop(maxTextureSize, [&](const IntRect& tileRect) {
+            RefPtr<BitmapTexture> surface = options.textureMapper.acquireTextureFromPool(tileRect.size(), { BitmapTexture::Flags::SupportsAlpha });
+            {
+                SetForScope<RefPtr<BitmapTexture>> scopedSurface(options.surface, surface);
+                SetForScope scopedOffset(options.offset, -toIntSize(tileRect.location()));
+                SetForScope scopedOpacity(options.opacity, 1);
+
+                options.textureMapper.bindSurface(options.surface.get());
+                paintFunction(options);
+
+                // If paintFunction applies filters to flattened surface then surface object might have
+                // changed to new intermediate filter surface and it needs to be stored for painting.
+                ASSERT(options.surface);
+                surface = options.surface;
+            }
+            m_textures.append(WTFMove(surface));
+        });
+
+        options.textureMapper.bindSurface(options.surface.get());
+        options.textureMapper.setDepthRange(prevZNear, prevZFar);
+    }
+
+    void paintToTextureMapper(TextureMapper& textureMapper, const FloatRect& targetRect, TransformationMatrix& modelViewMatrix, float opacity)
+    {
+        IntSize maxTextureSize = textureMapper.maxTextureSize();
+        auto allEdgesExposed = m_rect.width() <= maxTextureSize.width() && m_rect.height() <= maxTextureSize.height()
+            ? TextureMapper::AllEdgesExposed::Yes
+            : TextureMapper::AllEdgesExposed::No;
+        TransformationMatrix adjustedTransform = modelViewMatrix;
```

Comment:
You could reduce that back into one line and directly use operator *.

Reply:
Fixed